From 9ecae42cd0c19cf403641bbf9bf6a68b70b07344 Mon Sep 17 00:00:00 2001
From: Varun Kumar Reddy Yaparla <vyaparla@habana.ai>
Date: Mon, 13 Dec 2021 12:11:24 +0200
Subject: [PATCH] [SW-0] crc32: move crc32 table init to crc32 module

This change is required for boot info to
calculate crc while reading last boot info. Since boot
info reads are happenings before crc32 table initialization,
this change moves crc32 table initialization to crc32
module. crc32 table will be initialized once whenever
a crc32 call is made.

Change-Id: I9388be25d793e0c1895be85fee47399e465c5b50
---

diff --git a/app/preboot/preboot.c b/app/preboot/preboot.c
index a739925..7315b93 100644
--- a/app/preboot/preboot.c
+++ b/app/preboot/preboot.c
@@ -465,7 +465,7 @@
 
 	hl_ram_init();
 	preboot_dev_enable_caps();
-	crc32_init_lookup();
+
 	preboot_process_last_boot_info();
 	preboot_process_last_rma_info();
 
diff --git a/lib/preboot/crc32.c b/lib/preboot/crc32.c
index 3f85224..c999a1d 100644
--- a/lib/preboot/crc32.c
+++ b/lib/preboot/crc32.c
@@ -20,6 +20,7 @@
 #define TABLE_SIZE	256
 
 static u32 crc32table[256];
+static bool is_crc32_table_initialized;
 
 void crc32_init_lookup(void)
 {
@@ -43,6 +44,11 @@
 {
 	crc = ~crc;
 
+	if (!is_crc32_table_initialized) {
+		crc32_init_lookup();
+		is_crc32_table_initialized = true;
+	}
+
 	while (len--) {
 		crc ^= *p++;
 		crc = (crc >> 8) ^ crc32table[crc & 255];
